src/lib.rs

```diff
@@ -1,123 +1,441 @@
 
-        let verified_output = verify_quote_dcapv3(&dcap_quote, &collaterals, PINNED_TIME);
+    quote.signature.verify_qe_report()?;
 
-        println!("{:?}", verified_output);
-        let root_hash = hash_x509_keccak256(&collaterals.get_sgx_intel_root_ca());
-        let sign_hash = hash_x509_keccak256(&collaterals.get_sgx_tcb_signing());
-        let crl_hash = hash_crl_keccak256(&collaterals.get_sgx_intel_root_ca_crl().unwrap());
-        println!("{:?}", root_hash);
-        println!("{:?}", sign_hash);
-        println!("{:?}", crl_hash);
+    let mut key = [0u8; 65];
+    key[0] = 4;
+    key[1..].copy_from_slice(&quote.signature.attestation_pub_key);
+
+    if quote.header.attestation_key_type.get() != AttestationKeyType::Ecdsa256P256 as u16 {
+        bail!("unsupported attestation key type");
     }
 
-    #[test]
-    fn test_verifyv4() {
-        // let current_time = chrono::Utc::now().timestamp() as u64;
+    let attest_key = VerifyingKey::from_sec1_bytes(&key)
+        .map_err(|e| anyhow!("failed to parse attest key: {e}"))?;
 
-        let mut collaterals = IntelCollateral::new();
-        collaterals.set_tcbinfo_bytes(include_bytes!("../data/tcbinfov3_00806f050000.json"));
-        collaterals.set_qeidentity_bytes(include_bytes!("../data/qeidentityv2_apiv4.json"));
-        collaterals.set_intel_root_ca_der(include_bytes!("../data/Intel_SGX_Provisioning_Certification_RootCA.cer"));
-        collaterals.set_sgx_tcb_signing_pem(include_bytes!("../data/signing_cert.pem"));
-        collaterals.set_sgx_intel_root_ca_crl_der(include_bytes!("../data/intel_root_ca_crl.der"));
-        collaterals.set_sgx_platform_crl_der(include_bytes!("../data/pck_platform_crl.der"));
-        collaterals.set_sgx_processor_crl_der(include_bytes!("../data/pck_processor_crl.der"));
+    let header_bytes = quote.header.as_bytes();
+    let body_bytes = quote.body.as_bytes();
+    let mut data = Vec::with_capacity(header_bytes.len() + body_bytes.len());
+    data.extend_from_slice(header_bytes);
+    data.extend_from_slice(body_bytes);
 
+    let sig = quote.signature.isv_signature;
+    attest_key
+        .verify(&data, &sig)
+        .context("failed to verify quote signature")?;
 
-        let dcap_quote = QuoteV4::from_bytes(include_bytes!("../data/quote_tdx_00806f050000.dat"));
+    Ok(())
+}
 
-        let verified_output = verify_quote_dcapv4(&dcap_quote, &collaterals, PINNED_TIME);
+/// Ensure the latest tcb info is not revoked, and is either up to date or only needs a configuration
+/// change.
+fn verify_tcb_status(
+    current_time: SystemTime,
+    tcb_info: &TcbInfo,
+    pck_extension: &SgxPckExtension,
+) -> anyhow::Result<(TcbStatus, Vec<String>)> {
+    // Make sure current time is between issue_date and next_update
+    let current_time: DateTime<Utc> = current_time.into();
+    if current_time < tcb_info.issue_date || current_time > tcb_info.next_update {
```

Comment:
Wouldn't it be more efficient to verify expiration in `.as_tcb_info_and_verify` as part of `verify_integrity`? This way, it detects expired TCBInfo early, and wouldn't have to get into this step to begin with.

Reply:
Done ! Makes sense !